src/lib/Libutil/misc_utils.c

```diff
@@ -745,30 +746,28 @@ escape_delimiter(char *str, char *delim, char esc)
 		return strdup((char *)str);
 	}
 	delim_len = strlen(delim);
-	retstr = (char *)malloc(MAXPATHLEN);
-	if(retstr == NULL)
+	retstr = (char *) malloc(MAXBUFLEN);
+	if (retstr == NULL)
 		return NULL;
-	retstrlen = MAXPATHLEN;
+	retstrlen = MAXBUFLEN;
 
 	while (*str != '\0') {
-		if(*str == esc) {
+		if ((*str == esc && !(*(str + 1) == '\'' || *(str + 1) == '\"'))
+				|| (delim_len == 1 && *str == *delim)) {
 			retstr[i++] = esc;
+			retstr[i++] = *str++;
+		} else if (strncmp(str, delim, delim_len) == 0 && ((i + 1 + delim_len)
+				< retstrlen)) {
 			retstr[i++] = esc;
-				str++;
-		} else if(strncmp(str, delim, delim_len) == 0) {
-			retstr[i++] = esc;
-			for (j=0; j < delim_len; j++, i++) {
-				retstr[i] = delim[j];
-			}
-			str += delim_len;
-		} else {
+			for (j = 0; j < delim_len; j++, i++)
+				retstr[i] = *str++;
+		} else if ((i + 1 + delim_len) < retstrlen)
```

Comment:
This block shouldn't be executed if the same condition in line 759 is wrong but strncmp returns true since that might result in copying without adding escape character.